src/helpers/utilities.ts

```diff
@@ -106,16 +110,64 @@ export const parseAssetParam = (asset: string | undefined, api: ApiPromise): Mul
 };
 
 export const isPoolEmpty = (poolReserves: PoolReserves | undefined): boolean => {
-  return poolReserves?.[0] === BN_ZERO && poolReserves?.[1] === BN_ZERO;
+  return !poolReserves || poolReserves?.[0] === BN_ZERO || poolReserves?.[1] === BN_ZERO;
+};
+
+export const calcExchangeRate = (
+  poolReserves: PoolReserves,
+  asset1Decimals: number,
+  asset2Decimals: number,
+): Decimal => {
+  return new Decimal(unitToPlanck('1', asset1Decimals))
+    .mul(poolReserves[1].toString())
+    .divToInt(poolReserves[0].toString())
+    .div(unitToPlanck('1', asset2Decimals));
+};
+
+export const calcSwapAmountOut = (
+  poolReserves: PoolReserves,
+  amountIn: BN,
+  assetOutDecimals: number,
+  fee: number,
+): Decimal => {
+  const amountInWithFee = amountIn.mul(new BN(1000 - fee));
+  const numerator = amountInWithFee.mul(poolReserves[1]);
+  const denominator = poolReserves[0].mul(new BN(1000)).add(amountInWithFee);
+  return new Decimal(numerator.toString()).divToInt(denominator.toString()).div(unitToPlanck('1', assetOutDecimals));
+};
+
+export const calcSwapAmountIn = (
+  poolReserves: PoolReserves,
+  amountOut: BN,
+  assetInDecimals: number,
+  fee: number,
+): Decimal | null => {
+  if (amountOut.gte(poolReserves[1])) return null;
+
+  const numerator = poolReserves[0].mul(amountOut).mul(new BN(1000));
+  const denominator = poolReserves[1].sub(amountOut).mul(new BN(1000 - fee));
+  return new Decimal(numerator.toString())
+    .divToInt(denominator.toString())
+    .add(1)
+    .div(unitToPlanck('1', assetInDecimals));
+};
+
+const MAX_DECIMAL_POINTS = 5;
+export const formatDecimals = (value: Decimal, rounding: Decimal.Rounding = Decimal.ROUND_UP): string => {
+  return value.toDP(MAX_DECIMAL_POINTS, Decimal.ROUND_DOWN).toSignificantDigits(6, rounding).toString();
+};
+
+export const formatExchangeRate = (value: Decimal): string => {
+  const formatted = formatDecimals(value);
+  return value.gt(0) && formatted === '0' ? `<0.${'0'.repeat(MAX_DECIMAL_POINTS - 1)}1` : formatted;
 };
 
-export const calcExchangeRate = (amount1: number, amount2: number): Decimal | null => {
-  if (amount1 === 0 || amount2 === 0) return null;
-  return new Decimal(amount2).div(new Decimal(amount1));
+export const calcPriceImpact = (currentRate: Decimal, idealRate: Decimal): number => {
+  return new Decimal(100).sub(currentRate.mul(100).div(idealRate)).toDP(2).toNumber();
 };
 
-export const formatDecimals = (value: Decimal): string => {
-  return value.toSignificantDigits(6, Decimal.ROUND_UP).toString();
+export const calcSlippage = (amount: string, slippage: number, decimals: number): string => {
```

Comment:
lets separate into two functions
applyLesserSlippage
applyGreaterSlippage

Reply:
changed to applySlippage